lib/internal/modules/esm/get_format.js

```diff
@@ -82,6 +84,24 @@ function underNodeModules(url) {
   return StringPrototypeIncludes(url.pathname, '/node_modules/');
 }
 
+/**
+ * Determine whether the given source contains CJS or ESM module syntax.
+ * @param {string} source
+ * @param {URL} url
+ */
+function detectModuleFormat(source, url) {
+  try {
+    let realSource = source ?? readFileSync(url, 'utf8');
```

Comment:
An earlier version of `containsModuleSyntax` expected a file path, where it would load a file from disk and then parse it all in C++ land. Doing `readFileSync` here means that we cross from JS to C++ to get the file contents and send that large string back across the boundary, only to then send it right back again from JS to C++ for `containsModuleSyntax` to work with it.

A more performant approach is probably to update `containsModuleSyntax` to allow for an undefined input in the `source` parameter, and another optional parameter that can be a file path. Then in this situation where the JS side doesn’t already have the source, we can tell the C++ side to both read it from disk and analyze it, without multiple trips crossing the boundary. cc @anonrig

Reply:
I reintroduced the reading from file in the native code implementation for the case where `code` is not passed from the managed part. Slightly modified the logic as proposed [in this comment](https://github.com/nodejs/node/pull/53044#discussion_r1607875664).

The transpile case is currently the only one for which `resolve` cannot determine the format correctly. For that we only have the needed information during the load phase, not during the resolve hook is running.